### PR TITLE
[RFC] PVS static analysis fixes

### PR DESCRIPTION
**Pull request by peskal**
Fixes from PVS:
```
./src/nvim/api/private/helpers.c:763:1: warning: V560 A part of conditional expression is always false.
./src/nvim/api/private/helpers.c:764:1: warning: V560 A part of conditional expression is always false.
./src/nvim/digraph.c:1714:1: warning: V782 There is no sense in evaluating the distance between elements from different arrays: 'p - buf'.
./src/nvim/eval.c:2209:1: warning: V547 Expression 'key == NULL' is always false.
./src/nvim/eval.c:5709:1: warning: V547 Expression 'arg == NULL' is always false.
./src/nvim/eval.c:5836:1: warning: V547 Expression 'pt == NULL' is always false.
./src/nvim/eval.c:6225:1: warning: V547 Expression 'fname == NULL' is always false.
./src/nvim/eval.c:6312:1: warning: V547 Expression 'name == NULL' is always false.
./src/nvim/eval.c:7736:1: warning: V560 A part of conditional expression is always false: name == NULL.
./src/nvim/eval.c:8752:1: warning: V560 A part of conditional expression is always true: dashes != NULL.
./src/nvim/eval.c:8877:1: warning: V547 Expression 'name != NULL' is always true.
./src/nvim/eval.c:8930:1: warning: V547 Expression 'pt->pt_argv == NULL' is always false.
./src/nvim/eval.c:9200:1: warning: V547 Expression 'd != NULL' is always true.
./src/nvim/eval.c:9571:1: warning: V547 Expression 'pat != NULL' is always true.
./src/nvim/eval.c:10139:1: warning: V547 Expression 'd != NULL' is always true.
./src/nvim/eval.c:10243:1: warning: V547 Expression 'd != NULL' is always true.
./src/nvim/eval.c:14806:1: warning: V547 Expression 's == NULL' is always false.
./src/nvim/eval.c:18657:1: warning: V547 Expression 'pt->pt_argv == NULL' is always false.
./src/nvim/eval.c:18632:1: warning: V547 Expression 'pt != NULL' is always true.
./src/nvim/menu.c:1420:1: warning: V560 A part of conditional expression is always true: idx != - 1.
./src/nvim/move.c:1992:1: warning: V560 A part of conditional expression is always false: -- n < 0.
./src/nvim/move.c:2070:1: warning: V560 A part of conditional expression is always false: -- n < 0.
./src/nvim/tui/input.c:212:1: error: V547 Expression 'ev == TERMKEY_MOUSE_RELEASE' is always true.
./src/nvim/tui/tui.c:1679:1: warning: V782 There is no sense in evaluating the distance between elements from different arrays: 'bufp - bufs'.
```

### CODE REVIEW

File: src/nvim/tui/input.c [hunk 1]
```diff
@@ -209,7 +209,8 @@ static void forward_mouse_event(TermInput *input, TermKeyKey *key)
     }
   } else if (ev == TERMKEY_MOUSE_DRAG) {
     len += (size_t)snprintf(buf + len, sizeof(buf) - len, "Drag");
-  } else if (ev == TERMKEY_MOUSE_RELEASE) {
+  } else {
+    assert(ev == TERMKEY_MOUSE_RELEASE);
```
Comment (ZyX-I) [inline src/nvim/tui/input.c]: I would rather make that `switch()` with `case TERMKEY_MOUSE_UNKNOWN: { assert(false); }` and other cases from this if()/elseif()/… At least this way we will have a warning here if termkey adds more mouse events.

Comment (peskal) [reply src/nvim/tui/input.c]: I like that idea, the way I have it now seems a bit off. Only problem I see is that on line 174 it's already been verified that `ev` can only be `TERMKEY_MOUSE_RELEASE`, `TERMKEY_MOUSE_DRAG`, or `TERMKEY_MOUSE_PRESS`. If I change it to a switch statement with `case TERMKEY_MOUSE_UNKNOWN: { assert(false); }` PVS will probably complain about unreachable code.

Comment (peskal) [reply src/nvim/tui/input.c]: Nevermind, PVS is okay with the `switch`. The `default` statement is still unreachable though, that okay?

File: src/nvim/tui/input.c [hunk 2]
```diff
@@ -199,18 +199,24 @@ static void forward_mouse_event(TermInput *input, TermKeyKey *key)
...
-  } else if (ev == TERMKEY_MOUSE_DRAG) {
+    break;
+  case TERMKEY_MOUSE_DRAG:
     len += (size_t)snprintf(buf + len, sizeof(buf) - len, "Drag");
-  } else if (ev == TERMKEY_MOUSE_RELEASE) {
+    break;
+  case TERMKEY_MOUSE_RELEASE:
     len += (size_t)snprintf(buf + len, sizeof(buf) - len, "Release");
+    break;
+  default:
```
Comment (ZyX-I) [inline src/nvim/tui/input.c]: Don’t use `default` for enums. And `case`s need to be indented, they are only not indented in legacy code.

**Review by ZyX-I — CHANGES_REQUESTED** [review]

File: src/nvim/tui/input.c (hunk 1 above)
Comment (ZyX-I) [reply src/nvim/tui/input.c]: No `default` for enums, when enums happen to receive new value such switch()es need to be reviewed which will only happen without default.

Comment (peskal) [reply src/nvim/tui/input.c]: Gotcha, makes sense now

File: src/nvim/api/private/helpers.c [hunk 3]
```diff
@@ -760,12 +761,7 @@ bool object_to_vim(Object obj, typval_T *tv, Error *err)
     case kObjectTypeWindow:
     case kObjectTypeTabpage:
     case kObjectTypeInteger:
-      if (obj.data.integer > VARNUMBER_MAX
```
Comment (peskal) [inline src/nvim/api/private/helpers.c]: `VARNUMBER_MAX/MIN` is defined as maximum/minimum possible value of `varnumber_T` variable. `obj.data.integer` and `varnumber_T` are both defined as `int64_t`

File: src/nvim/eval.c [hunk 4]
```diff
@@ -5833,10 +5825,6 @@ static int get_lambda_tv(char_u **arg, typval_T *rettv, bool evaluate)
 
     fp = (ufunc_T *)xcalloc(1, sizeof(ufunc_T) + STRLEN(name));
     pt = (partial_T *)xcalloc(1, sizeof(partial_T));
-    if (pt == NULL) {
```
Comment (justinmk) [inline src/nvim/eval.c]: Really cool that PVS is catching these never-NULL `x*alloc` results. No other static analyzer has done that.

**Comment by justinmk** [comment]
LGTM. @ZyX-I ok?

File: src/nvim/api/private/helpers.c [hunk 5]
```diff
@@ -760,12 +761,7 @@ bool object_to_vim(Object obj, typval_T *tv, Error *err)
...
-          || obj.data.integer < VARNUMBER_MIN) {
-        api_set_error(err, kErrorTypeValidation, "Integer value outside range");
-        return false;
-      }
-
+      STATIC_ASSERT(sizeof(obj.data.integer) <= sizeof(varnumber_T), "Expected integer size to be less than or equal to VimL number size");
```
Comment (ZyX-I) [inline src/nvim/api/private/helpers.c]: Looks like too long line. Below there also is `--n;` which is not going to pass the linter.

File: src/nvim/tui/input.c [hunk 6]
```diff
@@ -199,18 +199,25 @@ static void forward_mouse_event(TermInput *input, TermKeyKey *key)
...
+      break;
+    case TERMKEY_MOUSE_DRAG:
+      len += (size_t)snprintf(buf + len, sizeof(buf) - len, "Drag");
+      break;
+    case TERMKEY_MOUSE_RELEASE:
+      len += (size_t)snprintf(buf + len, sizeof(buf) - len, "Release");
+      break;
+    case TERMKEY_MOUSE_UNKNOWN:
+      assert(false);
+      break;
```
Comment (ZyX-I) [inline src/nvim/tui/input.c]: LGTM, except for the three linter issues and that normally `assert(false)` goes without `break`.

**Review by ZyX-I — CHANGES_REQUESTED** [review]

**Comment by peskal** [comment]
@ZyX-I fixed! Btw, I ran the linter again and it returned a bunch of errors in the files I edited, but not around places I edited. Tried digging into the Makefiles but couldn't figure out the reason it works this way. This normal? Here's the errors:

```
src/nvim/eval.c:8732:  /*-style comment found, it should be replaced with //-style.  /*-style comments are only allowed inside macros.  Note that you should not use /*-style comments to document macros itself, use doxygen-style comments for this.  [readability/old_style_comment] [5]
src/nvim/eval.c:8733:  Do not use preincrement in statements, including for(;; action)  [readability/increment] [4]
src/nvim/menu.c:1422:  /*-style comment found, it should be replaced with //-style.  /*-style comments are only allowed inside macros.  Note that you should not use /*-style comments to document macros itself, use doxygen-style comments for this.  [readability/old_style_comment] [5]
src/nvim/menu.c:1424:  if should always use braces  [readability/braces] [5]
src/nvim/menu.c:1426:  Inner expression should be aligned as opening brace + 1 (+ 2 in case of {)  [whitespace/alignment] [2]
src/nvim/menu.c:1428:  An else should appear on the same line as the preceding }  [whitespace/newline] [4]
src/nvim/menu.c:1427:  If an else has a brace on one side, it should have it on both  [readability/braces] [5]
src/nvim/menu.c:1429:  Inner expression should be aligned as opening brace + 1 (+ 2 in case of {)  [whitespace/alignment] [2]
src/nvim/menu.c:1429:  Use true instead of TRUE.  [readability/bool] [4]
src/nvim/menu.c:1431:  If an else has a brace on one side, it should have it on both  [readability/braces] [5]
src/nvim/move.c:1993:  Do not use preincrement in statements, use postincrement instead  [readability/increment] [5]
src/nvim/move.c:2070:  Do not use preincrement in statements, use postincrement instead  [readability/increment] [5]
src/nvim/tui/input.c:207:  Lines should be <= 80 characters long  [whitespace/line_length] [2]

```

I fixed them locally and can push if needed. Funny thing is that it complains about things like /*-style comments in one place, but not in another. Or it complains about preincrement on one line, but not the line immediately after.

**Comment by ZyX-I** [comment]
@peskal This is a sign that you need to merge or rebase: these places should’ve been edited in master. Do not fix them befor that, this will spawn useless conflicts.

Linter complains about lines that were edited, otherwise there would be problems with our large non-style-compliant codebase. But the error suppression file is generated by a travis “cron job” bot and saved to github, not on your machine from your master branch.

**Comment by peskal** [comment]
@ZyX-I thanks, that makes sense that you'd suppress lines that weren't changed. I'd like to add this info to `test/README.md`, but I think I must not fully understand yet because I'm still having trouble getting the linter to work. I rebased onto master but I'm still getting these linter errors.

I did:
```
git fetch origin master
git rebase origin/master
```
then `git log` to verify that `origin/master`'s HEAD is my branch's merge base. Also `git merge-base master static-analysis-fixes` returns master's HEAD at this moment: `9393be4`

However `make lint` still fails with the error:
```
[  0%] Generating ../../touches/ran-clint-eval-c
src/nvim/eval.c:8732:  /*-style comment found, it should be replaced with //-style.  /*-style comments are only allowed inside macros.  Note that you should not use /*-style comments to document macros itself, use doxygen-style comments for this.  [readability/old_style_comment] [5]
Total errors found: 1
```

You said
> the error suppression file is generated by a travis “cron job” bot and saved to github, not on your machine from your master branch.

Is there more documentation somewhere on how the travis build works? How can I lint my changes locally before pushing?

**Comment by ZyX-I** [comment]
@peskal Well, you edited the adjacent line, this automatically disabled suppression. Partially because this is way too easier to program suppressions this way (two errors from different builds are considered to be the same if their 5-tuples (file, error itself, line before error, line with error, line after error) are the same: you may see how easy this is to check), partially because this forces contributors to do style changes when they alter adjacent code which slowly makes code base cleaner. CONTRIBUTING.md has a word about this behaviour.

**Comment by ZyX-I** [comment]
And note that bot makes updates to suppression database at cron jobs, not at pushes. Most of time you would not hit any issues due to that though: newer changes should have less linter failures, not more. Rare exceptions are updates to the linter itself and a few cases where lint issues were somehow overlooked or delibarately ignored with intent to make bot silence the issues (though I remember at least one case when I added commit with fixes to that in some of my PRs before bot updated database).

**Comment by justinmk** [comment]
Fixed lint errors in merge. Thanks @peskal and @ZyX-I !

**Comment by peskal** [comment]
Thanks @justinmk. Looks like the lint errors you fixed were the exact ones I was seeing as well. Ok, so it sounds to me like it's good practice to run `make lint` before each commit. If I do that, it should mean less work for you during merge.

**Pull request merged**